### PR TITLE
use min_specialization for some rustc crates where it requires no changes

### PR DESCRIPTION
**Pull request by RalfJung**
and add FIXME for the rest

Cc @matthewjasper

### CODE REVIEW

File: src/librustc_metadata/lib.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@
 #![feature(nll)]
 #![feature(or_patterns)]
 #![feature(proc_macro_internals)]
-#![feature(specialization)]
+#![feature(specialization)] // FIXME: min_specialization ICEs
```
Comment (RalfJung) [inline src/librustc_metadata/lib.rs:10]: @matthewjasper ICEs shouldn't happen, should they?
```
error: internal compiler error: src/librustc_trait_selection/traits/specialize/mod.rs:101: When translating substitutions for specialization, the expected specialization failed to hold

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:904:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Comment (matthewjasper) [reply src/librustc_metadata/lib.rs:10]: They shouldn't, but the feature has pretty limited testing.

**Comment by matthewjasper** [comment]
Can you test `libserialize` as well?

**Comment by RalfJung** [comment]
Ah I only grepped in `libr*`and missed that one... on it.

**Comment by RalfJung** [comment]
Nope that did not work.
```
error: cannot specialize on trait `serialize::Encodable`
   --> src/libserialize/serialize.rs:644:1
    |
644 | / impl<T: ?Sized + Encodable> Encodable for Box<T> {
645 | |     fn encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
646 | |         (**self).encode(s)
647 | |     }
648 | | }
    | |_^

error: cannot specialize on trait `serialize::Decodable`
   --> src/libserialize/serialize.rs:650:1
    |
650 | / impl<T: Decodable> Decodable for Box<T> {
651 | |     fn decode<D: Decoder>(d: &mut D) -> Result<Box<T>, D::Error> {
652 | |         Ok(box Decodable::decode(d)?)
653 | |     }
654 | | }
    | |_^
```
And if I mark those as specializable then tons of other impls fail.

**Comment by RalfJung** [comment]
rebased
@bors r=matthewjasper

**Pull request merged**